apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -475,6 +475,10 @@ public void deactivate(TraceContextHolder<?> holder) {
         }
         final Deque<TraceContextHolder<?>> stack = activeStack.get();
         assertIsActive(holder, stack.poll());
+        if (!stack.isEmpty() && !holder.isDiscard()) {
+            //noinspection ConstantConditions
+            stack.peek().setDiscard(false);
```

Comment:
don't we have to set discard to false on all elements in the stack?

Reply:
Not really. All spans are created as non-discardable, except from the `trace_methods` spans. This means there is no race between `deactivate` and `end` and it should be safe to do it this way.

Reply:
gotcha